docs/rfc/rfc-6.adoc

```diff
@@ -0,0 +1,228 @@
+
+The first member responsible for submitting the DKG result is a member with
+index `hash(new_group_pubkey) % group_size`. If that member did not submit the
+result,  `(hash(new_group_pubkey) % group_size) + 1` becomes responsible next.
+Group members who were inactive and skipped their responsibility might be
+reported by the wallet as ones who failed the heartbeat and marked as ineligible
+for rewards.
+
+For example, if `hash(new_group_pubkey) % group_size = 62`, `group_size = 64`,
+and member `9` submitted DKG result, members with indexes
+`{62, 63, 64, 1, 2, 3, 4, 5, 6, 7, 8}` might be marked as ineligible for
+rewards. Inactive group members are determined once the result is submitted and
+accepted by the chain.
+
+Given that the DKG result submission process consists of two transactions:
+submitting the result and approving the result after the challenge period ends,
+in order for the operator to be considered active, they must submit **both**
+transactions.
+
+===== Report heartbeat failed
+
+A wallet should periodically execute off-chain heartbeats and report inactive
+operators responsible for potential heartbeat failures. The majority of wallet
+members need to come together and sign a message with their private ECDSA
+operator keys. The message should contain the list of inactive operators and
+nonce ensuring uniqueness. The same mechanism can be used for reporting
+operators who failed their duty of submitting DKG result. 
+
+No submission order will be enforced on-chain for the heartbeat failure
+transaction and the transaction submitter will have ETH cost reimbursed. To
+disincentivize front-running, no additional reward will be released by this
+transaction. Off-chain clients should have a handshake agreement to follow the
+submission order: the operator submitting this TX is the first active one.
+
+When reporting heartbeat failure as a result of DKG result submission
+inactivity, the group must decide if the cost of the transaction is worth it.
+That is, if some operator was marked as inactive a minute ago, it does not make
+sense to mark it as inactive and ineligible for rewards one more time.
+
+IMPORTANT: There should be no token incentives for calling this function by
+the group. The only incentive should be their goodwill to save some rewards by
+DAO. Moreover, in the case of off-chain heartbeat failure, clients are trying to
+avoid slashing in the future. Marking operators as ineligible for rewards should
+not increase the rewards of other group members! Rewards saved on inactive
+operators marked as ineligible should return back to the DAO. The DAO should do
+something else with those extra tokens other than increase rewards. We do not
+want to incentivize voting for your own salary.
+
+[[maintainer-incentives]]
+==== Maintainer incentives
+
+For the first release, the role of maintainers will belong to the willing
+off-chain clients approved by DAO or council. Every client will have maintainer
```

Comment:
I think that the `Bridge` contract should not contain any logic related to reimbursements. This should all happen in the `Maintainer` contract proxying calls. If we decide to move to Gelato, we will just remove `Maintainer` and have Gelato network calling `Bridge` directly.

If someone wants to bypass `Maintainer`, fine. They will pay for the transaction and don't get reimbursed. An important part is to make sure that in this case, all `Bridge` functions revert early to do not burn too much gas from frontrunned honest maintainer.

Yes, maintainers are approved on the `Maintainer` contract. They are not trusted anymore than 3rd party anons calling `Bridge` functions directly. The only reason to approve them is to make sure they follow the off-chain agreement and do not frontrun each other.

Reply:
> The only reason to approve them is to make sure they follow the off-chain agreement and do not frontrun each other.

What kind of validations of the client are supposed to be done by the `Maintainer` contract or DAO/council members when approving a client? I'm not familiar with how it works - will DAO somehow have access to the code of the client and will be able to verify that client was not modified to skip the order of transaction submission?
Or maybe this ensuring that off-chain order is followed will be done retroactively - DAO first approves a client and later if that client misbehaves (e.g. skips the line), DAO can step in and remove that client from the list of approved maintainers?

Reply:
The entire idea of the approval is just to have a smaller group of people submitting maintainer transactions so that we are sure that this group is not modifying their clients to race with each other. There is no formal process. One approach could be e.g. having the bootstrap nodes to be the maintainers plus some group of other clients.